youtube_dl/downloader/__init__.py

```diff
@@ -30,6 +30,10 @@
 def get_suitable_downloader(info_dict, params={}):
     """Get the downloader class that can handle the info dict."""
     protocol = determine_protocol(info_dict)
+    if protocol == 'm3u8' and params.get('hls_prefer_native') is True:
+        protocol = 'm3u8_native'
+    elif protocol == 'm3u8_native' and params.get('hls_prefer_native') is False:
+        protocol = 'm3u8'
```

Comment:
I don't see the problem,
we change protocol before storing it to info_dict['protocol'].

There's two advantages:
-We fix the current problem whith hls-prefer-ffmpeg without breaking anything.
-We have a cleaner code without multiple returns.